File: NeetCode/Arrays_and_Hashing/longest_consecutive_sequence/longest_consecutive_sequence.py
from typing import Optional, List

import heapq


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        s = set(nums)
        best = 0

        for num in s:
            if num - 1 not in s:
                y = num
                while y in s:
                    y += 1
                best = max(best, y - num)
        return best


if __name__ == "__main__":
    sol = Solution()

    nums = [0, 3, 7, 2, 5, 8, 4, 6, 0, 1]
    print(sol.longestConsecutive(nums))
    print("Running Solution...")

File: NeetCode/Arrays_and_Hashing/longest_consecutive_sequence/longest_consecutive_sequence.cpp
// longest_consecutive_sequence.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int longestConsecutive(vector<int> &nums)
    {
        int best = 0;
        unordered_set<int> values(nums.begin(), nums.end());

        for (int val : values)
        {
            if (values.count(val - 1) > 0)
            {
                continue;
            }
            int nxt = val;
            while (values.count(nxt) > 0)
            {
                nxt += 1;
            }
            best = max(best, nxt - val);
        }
        return best;
    }
};